### ISSUE TITLE
Design for Maptip Support of Feature Based Layers

### ISSUE THREAD

**Issue by mweech** [fgpv-vpgf/fgpv-vpgf#1337]
Utilize the notes from #424 to complete design of the following maptip functionality:

- support anchor tooltips on feature click
- anchor tooltips will only work on client side based feature layers (not dynamic/image based layers)
- if map author selects to utilize the anchor tip mode of selection, the default identify behavior will be disabled (not intended for usage on maps with mix of image/feature based layers)
- initial output of the tooltip will follow same format as identify results
- considerations should be made for the possibility of adding tooltip templating support in the future

**Comment by james-rae** [comment]
## Things to do

- [ ] Add setting to config schema to enable or disable maptip mode
- [ ] Implement a physical maptip (box container, close button, name & icon content similar to identify list)
- [ ] Trigger maptip on feature click - position & content, close any existing tip
- [ ] Flag check at beginning of identify to abort if maptip mode (or something that prevents the event from registering)
- [ ] Handle repositioning or resizing of maptip if map extent changes
- [ ] Close maptip if layer removed or set to invisible
- [ ] Implement close button handler (hard!)

## Relevant Code and API Calls

[Get extent](https://developers.arcgis.com/javascript/3/jsapi/esri.graphicsutils-amd.html#graphicsextent) of a graphic.  Will be in map co-ordinates, if need to get the top-middle in screen co-ordinates, use [toScreen function](https://developers.arcgis.com/javascript/3/jsapi/map-amd.html#toscreen)

Alternate approach of [getting screen extent of graphic using SNAP library](https://github.com/RAMP-PCAR/RAMP-PCAR/blob/v5.5.0-10/src/js/RAMP/Modules/featureHighlighter.js#L260)

ESRI featurelayer [mouse click event](https://developers.arcgis.com/javascript/3/jsapi/featurelayer-amd.html#event-click). We would be interested in the `graphic` property of the event parameter.  The graphic object has a `getLayer()` function to find out who it belongs to.

RAMP 1 triggers the anchor tip when a highlighted feature is [added to the highlight layer](https://github.com/RAMP-PCAR/RAMP-PCAR/blob/v5.5.0-10/src/js/RAMP/Modules/featureHighlighter.js#L297). A bit backwards to how we will probably do it but linking anyways.

## Outstanding Details

Best way to make the physical maptip.  See issue sub-section [Creating a Physical Maptip](https://github.com/fgpv-vpgf/fgpv-vpgf/issues/424#issuecomment-221270498). 
TL;DR current options are 
* homemade Angular tip
* 3rd party library (e.g. Tooltipster)
* SVG Tooltip within a Graphic Layer


## Potential Additional Features

* Button to trigger Show Details?
* Highlighting of clicked feature?
* Fast Tooltip? -- as the app is architected right now, if a user hits a tooltip prior to opening a grid, it will trigger the entire attribute set of a layer to download (and the tooltip won't show until it finishes).  We could attempt to fancy things up by testing if the attributes have not been downloaded, and in that case hit the server for just the one feature.  Could utilize geoApi function [getFeatureInfo](https://github.com/fgpv-vpgf/geoApi/blob/v1.1.0-3/src/layer.js#L1007). Remember that the local in-layer graphics should only have geometry and object id attribute.

**Comment by AleksueiR** [comment]
#### Some thoughts in no particular order:

- If tooltips will only be supported on client-side Feature layers, why do we limit ourselves to `click` maptips? We can easily show maptips on `hover` as well.
- Will the maptip show only a single feature or multipe features (JOSM style)?
![](https://i.imgur.com/HXzXZYO.png)
![](https://i.imgur.com/5qjWssV.png)
- What are use cases for this or proposed usage scenarios?
- If maptip is going to display multiple results with buttons that trigger Show Details, then, in essence, it's an identify click with some extra text attached to the identify anchor. Is this maptip going to replace the identify panel?
- Fast Tooltip - yes, I think we need to do both: hit the server for just one record and trigger the attribute download. If the user clicked on one thing, it's likely he will click on something else, so it's better to start preloading data at this point.
- SNAP library hasn't been updated for some time now. It's better to use SVG.JS
- I think creating an svg-based maptip and piggy-backing it onto the highlight svg layer to make it move when the map is panned (might need to recreate it when the map is zoomed - could be a scaling issue here)
